### PR TITLE
TSan instrumentation of module code

### PR DESCRIPTION
**Pull request by bjacob**
This implements TSan instrumentation in the IREE compiler, and adds a `IREE_BYTECODE_MODULE_ENABLE_TSAN` flag, which when set causes `iree_bytecode_module` to enable TSan instrumentation. 

When `IREE_BUILD_TESTS` is on, we enforce early that `IREE_ENABLE_TSAN` and `IREE_BYTECODE_MODULE_ENABLE_TSAN` agree, as we would otherwise get test crashes anyway.

`IREE_BYTECODE_MODULE_ENABLE_TSAN` also requires `IREE_BYTECODE_MODULE_FORCE_SYSTEM_DYLIB_LINKER` to be set. That too is enforced early.

Putting all that together, with this PR, the incantation to build tests with TSan is:
```
cmake \
  -DIREE_BUILD_TESTS=ON \
  -DIREE_ENABLE_TSAN=ON \
  -DIREE_BYTECODE_MODULE_ENABLE_TSAN=ON \
  -DIREE_BYTECODE_MODULE_FORCE_SYSTEM_DYLIB_LINKER=ON \
   ...other flags...

cmake --build .
cmake --build . --target iree-test-deps
```

So clearly this isn't optimal in the sense of minimizing required typing to enable TSan, but at least this is optimizing "options that do one clear thing each" and "reuse existing cmake logic (`IREE_BYTECODE_MODULE_FORCE_SYSTEM_DYLIB_LINKER`).

This PR is ending up simpler than anticipated because
* We used to have trouble with `iree_bytecode_module` setting the system-linker-path but that was fixed in #8591
* We used to have to handle switching to the system-linker as opposed to embedded-linker, but the cmake infra for that (`IREE_BYTECODE_MODULE_FORCE_SYSTEM_DYLIB_LINKER`) was added in #8659 .
* We used to run into lots of trouble with TSan apparently not supporting vectors with more than 16 elements, I first tackled that using LLVM Scalarizer pass but ran into a [bug](https://github.com/llvm/llvm-project/issues/54469), then I wrote another approach using VectorUnrollPatterns ([diff here](https://gist.github.com/bjacob/b0d967327ab53574d99d8ff9875f582c)), but in the end I noticed that none of this seemed to be needed anymore --- I want to believe that that's because something was changed in LLVM but I haven't been able to pinpoint it. Anyway, if we run into issues where some vector load/stores of more than 16 elements are not tsan-instrumented, we can revive that diff.

### CODE REVIEW

**Comment by benvanik** [comment]
There's no relation - you pass the flag when you compile a module if you want the module to have the sanitizer enabled. Building the compiler with a sanitizer does not imply the artifacts need to be (or that we want them to be - say if you are cross compiling) and building the runtime with a sanitizer does not imply the artifacts need to be either (or can be - you may be loading a file from someone else, etc). The only requirement is that if you build an artifact with a sanitizer then you must also build the runtime like that. So: enable the sanitizer for the runtime, and pass the sanitizer flag to the compiler when you are building things you want to be sanitized.

File: iree/compiler/Dialect/HAL/Target/LLVM/internal/UnixLinkerTool.cpp [hunk 1]
```diff
@@ -28,7 +28,11 @@ class UnixLinkerTool : public LinkerTool {
     if (!toolPath.empty()) return toolPath;
 
     // No explicit linker specified, search the environment for common tools.
-    toolPath = findToolInEnvironment({"ld", "ld.gold", "ld.lld"});
+    if (useClangDriver()) {
+      toolPath = findToolInEnvironment({"clang"});
+    } else {
+      toolPath = findToolInEnvironment({"ld", "ld.gold", "ld.lld"});
+    }
```
Comment (ScottTodd) [inline iree/compiler/Dialect/HAL/Target/LLVM/internal/UnixLinkerTool.cpp:35]: Here's another spot where I'd rather at least start with the explicit configuration then consider adding implicit switches.

If this can be achieved by setting the linker to clang explicitly (`-iree-llvm-system-linker-path=` flag, `IREE_LLVMAOT_SYSTEM_LINKER_PATH` env var), then we wouldn't need all this branching in this file.

File: build_tools/cmake/iree_bytecode_module.cmake [hunk 2]
```diff
@@ -114,6 +114,23 @@ function(iree_bytecode_module)
...
+  # time being) a host build dir, so any such tests are by definition build-only
+  # which makes it unnecessary to enable TSan in module code. We could support
+  # that if needed, it's just that for now, as TSan linking is implemented using
+  # clang as the linker driver, we would have to figure out the right (-march?)
+  # flag to pass for cross-linking.
+  if (IREE_ENABLE_TSAN AND NOT (_ARGS MATCHES "-iree-llvm-target-triple="))
+    list(APPEND _ARGS "-iree-llvm-sanitize=thread")
+    # Avoid undefined-symbol errors on tsan symbols e.g. __tsan_read8
+    list(APPEND _ARGS "-iree-llvm-link-embedded=false")
+  endif()
```
[the remark this answers is not in the document]
Comment (bjacob) [reply build_tools/cmake/iree_bytecode_module.cmake:132]: ooo today but this is the high level question that i'd need to understand better before going into impl details:  My goal here has been "Building and running tests (CTest) in a build dir with `IREE_ENABLE_TSAN=ON` should work".  Are you (1) agreeing with the goal but suggesting an alternate impl strategy or (2) suggesting an alternate goal?  (Either way please elaborate on the alternate suggestion, more than happy to implement whatever reviewers suggest - I just need something actionable).

[the remark this answers is not in the document]
Comment (ScottTodd) [reply build_tools/cmake/iree_bytecode_module.cmake:132]: Having a single setting that affects all builds/tests is ambitious, so I'd start more focused, based on priority and expected usage. Is there a particular part of the project you want TSAN coverage for?

1. Base runtime tests/binaries (cc_test, some of the HAL CTS, tools like generate_embed_data)
2. Compiler-adjacent runtime tests/binaries (check_test/iree-run-module not using CPU codegen)
3. Compiler tests/binaries (lit_test, iree-opt)
4. CPU compiled code (check_test/iree-run-module using LLVM CPU codegen)

It looks like you're jumping straight to 4 (the most complex), when excluding CPU generated code from TSAN would be a simpler way to start. We can exclude LLVM CPU codegen tests from a full "build+run all tests" config if we mainly want coverage for code outside of CPU codegen, for example by only enabling `IREE_TARGET_BACKEND_VMVX` and `IREE_HAL_DRIVER_VMVX`.

If we really do want coverage for the cpu codegen, then it's a question of whether we need to fully integrate it into the build system as you're close to here, or if just keeping support around for offline developer use is sufficient. The `iree-llvm-sanitize` flag has been in that state (not wired up to any CI) since it was added.

[the remark this answers is not in the document]
Comment (bjacob) [reply build_tools/cmake/iree_bytecode_module.cmake:132]: I really have this high level goal that one should be able to toggle `IREE_ENABLE_TSAN=ON`, rebuild, rerun tests, and that would do what one would naively think it does: the same tests are now running with TSan enabled. If I followed the approach outline in the previous comment, only enabling the VMVX driver/backend, then setting `IREE_ENABLE_TSAN=ON` would either (the current behavior) cause CPU codegen tests to fail, or (the only alternative) be skipped. We would be digging deeper the hole that we are already in with the complex matrix of cmake options that work / don't work with each other, and the complex set of tests that are expected to fail in each case.   I'm interested in fixing this high level problem that the IREE project currently has.

**Comment by bjacob** [comment]
Updates:
1. This is no longer using the `clang` driver.  All what was needed was passing `-shared` instead of `-static`.
2. In today's meeting we agreed that `IREE_ENABLE_TSAN` needs to be split into two settings, one enabling TSan in the compiler and the other enabling TSan in the runtime. Once that is done, the `iree_bytecode_module` logic debated above will just need to follow the runtime setting (which is where it's necessary, due to TSan breaking ABI) and that will thus become decoupled from the compiler. I will get back to this once https://reviews.llvm.org/D121182 is merged, and then, I will need to understand how exactly we implement that compiler/runtime decoupling --- based on source code directories or on individual cmake rules ?

File: build_tools/cmake/iree_bytecode_module.cmake [hunk 3]
```diff
@@ -114,6 +114,23 @@ function(iree_bytecode_module)
...
+  #
+  # We make an exception for cross-compiling tests, crudely identified for now
+  # by the presence of a target-triple flag. IREE_ENABLE_TSAN implies (for the
+  # time being) a host build dir, so any such tests are by definition build-only
+  # which makes it unnecessary to enable TSan in module code. We could support
+  # that if needed, it's just that for now, as TSan linking is implemented using
+  # clang as the linker driver, we would have to figure out the right (-march?)
+  # flag to pass for cross-linking.
+  if (IREE_ENABLE_TSAN AND NOT (_ARGS MATCHES "-iree-llvm-target-triple="))
+    list(APPEND _ARGS "-iree-llvm-sanitize=thread")
```
Comment (ScottTodd) [inline build_tools/cmake/iree_bytecode_module.cmake:129]: <sub>(adding as a comment on these lines so the conversation can be replied to directly and can be marked as "resolved", unlike top level PR comments)</sub>

> 2. In today's meeting we agreed that `IREE_ENABLE_TSAN` needs to be split into two settings, one enabling TSan in the compiler and the other enabling TSan in the runtime. Once that is done, the `iree_bytecode_module` logic debated above will just need to follow the runtime setting (which is where it's necessary, due to TSan breaking ABI) and that will thus become decoupled from the compiler. I will get back to this once https://reviews.llvm.org/D121182 is merged, and then, I will need to understand how exactly we implement that compiler/runtime decoupling --- based on source code directories or on individual cmake rules ?

How about keeping one set of options that passes `-fsanitize=[address,thread,memory]` to the CMake build system, affecting "runtime" binaries and "compiler" binaries: https://github.com/google/iree/blob/e9c786761f91f02e854875ca5b8b257cc706a993/build_tools/cmake/sanitizers.cmake#L16-L19

then adding a new set of options that adds `-iree-llvm-sanitize=[address,thread,memory]` to the CMake functions that invoke IREE's compiler.

Brainstorming names:
* `IREE_CODEGEN_ENABLE_TSAN` 
* `IREE_ENABLE_CODEGEN_TSAN`
* `IREE_TESTS_ENABLE_TSAN` 
* `IREE_ENABLE_TSAN_IN_TESTS`

Then sample flag usage:

Configuration | How to set flags
-- | --
Linux host + Linux target | Single CMake configure with both flags set
Linux host + Android target | Configure host with `IREE_ENABLE_TSAN_IN_TESTS`, target with `IREE_ENABLE_TSAN`

---

I don't think a split between "runtime binaries" and "compiler binaries" would make sense for this flag, like we have for the Tracing option, but here's how that is done:

https://github.com/google/iree/blob/e9c786761f91f02e854875ca5b8b257cc706a993/CMakeLists.txt#L35-L36

https://github.com/google/iree/blob/e9c786761f91f02e854875ca5b8b257cc706a993/iree/compiler/Utils/CMakeLists.txt#L58-L63

https://github.com/google/iree/blob/e9c786761f91f02e854875ca5b8b257cc706a993/iree/compiler/Utils/TracingUtils.h#L25-L30

File: iree/compiler/Dialect/HAL/Target/LLVM/internal/EmbeddedLinkerTool.cpp [hunk 4]
```diff
@@ -55,7 +55,7 @@ class EmbeddedLinkerTool : public LinkerTool {
 
     // No explicit linker specified, search the environment for common tools.
     std::string environmentPath =
-        findToolInEnvironment({"iree-lld", "lld", "ld.lld", "lld-link"});
+        findToolInEnvironment({"iree-lld", "ld.lld", "lld", "lld-link"});
```
Comment (ScottTodd) [inline iree/compiler/Dialect/HAL/Target/LLVM/internal/EmbeddedLinkerTool.cpp:58]: This looks unrelated to the rest of this PR.

File: iree/hal/local/loaders/system_library_loader.c [hunk 5]
```diff
@@ -160,6 +160,7 @@ static iree_status_t iree_hal_system_executable_query_library(
 
   // Ensure that if the library is built for a particular sanitizer that we also
   // were compiled with that sanitizer enabled.
+  fprintf(stderr, "header->sanitizer = %d\n", header->sanitizer);
```
Comment (ScottTodd) [inline iree/hal/local/loaders/system_library_loader.c:163]: (remove debug printing before merge)

File: build_tools/cmake/iree_bytecode_module.cmake [hunk 6]
```diff
@@ -108,13 +108,23 @@ function(iree_bytecode_module)
...
     list(APPEND _ARGS "-iree-llvm-embedded-linker-path=\"${_LINKER_TOOL_EXECUTABLE}\"")
   endif()
 
+  # Support testing in TSan build dirs. Unlike other sanitizers, TSan is an
+  # ABI break: when the host code is built with TSan, the module must be too,
+  # otherwise we get crashes calling module code.
+  if (IREE_ENABLE_TSAN_IN_TEST_MODULES)
+    list(APPEND _ARGS "-iree-llvm-sanitize=thread")
+    # Avoid undefined-symbol errors on tsan symbols e.g. __tsan_read8
+    list(APPEND _ARGS "-iree-llvm-link-embedded=false")
```
Comment (ScottTodd) [inline build_tools/cmake/iree_bytecode_module.cmake:125]: Let's add a bit more to this comment, since this is a behavior change.

```suggestion
    # TSan requires some imported symbols like __tsan_read8, so
    # we must use the system linker with shared (not static) linking.
    # This changes behavior for tests that were expecting to use
    # the embedded linker, but at least we get some coverage
    # using TSan this way.
    list(APPEND _ARGS "-iree-llvm-link-embedded=false")
```

**Review by ScottTodd — COMMENTED** [review]
Thanks, splitting the option helped, and I'm happy enough with this now (once some leftover debug code is cleaned up and https://github.com/google/iree/pull/8591 or some other fix for `lld` getting passed to `UnixLinkerTool` is merged).

File: CMakeLists.txt [hunk 7]
```diff
@@ -126,6 +126,26 @@ option(IREE_ENABLE_LLD "Use lld when linking" OFF)
 option(IREE_ENABLE_ASAN "Enable address sanitizer" OFF)
 option(IREE_ENABLE_MSAN "Enable memory sanitizer" OFF)
 option(IREE_ENABLE_TSAN "Enable thread sanitizer" OFF)
+option(IREE_BYTECODE_MODULE_ENABLE_TSAN "Enable thread sanitizer in IREE modules in tests" OFF)
+
+if(NOT IREE_ENABLE_TSAN STREQUAL IREE_BYTECODE_MODULE_ENABLE_TSAN)
```
[the remark this answers is not in the document]
Comment (GMNGeoffrey) [reply CMakeLists.txt:131]: I think it's also worth a brief comment here about why we're doing it this way (as opposed to having one flag).

[the remark this answers is not in the document]
Comment (bjacob) [reply CMakeLists.txt:131]: I dropped the `if(IREE_BUILD_TESTS)` because I realized that the actual relevant condition here is "are we building any `iree_bytecode_module`) and that's not quite the same? Would you know how to express that --- is this some OR combination of things like IREE_BUILD_TESTS, IREE_BUILD_SAMPLES, ... ?

Re `STREQUAL` - do you mean I should be able to use `EQUAL` here ?  I dont really understand that that does, sorry. Is it "equality comparison after implict conversion to bool" ?

**Review by MaheshRavishankar — APPROVED** [review]
Looks fine to me. I thought you needed vector unrolling. Dont see it here.

**Comment by bjacob** [comment]
> Looks fine to me. I thought you needed vector unrolling. Dont see it here.

See the updated PR description. The problem mysteriously vanished.

File: CMakeLists.txt (hunk 7 above)
[the remark this answers is not in the document]
Comment (GMNGeoffrey) [reply CMakeLists.txt:131]: I think it's really "Are we building and then executing any iree_bytecode_module". We could be building it for use on some other platform (or could be executing one we built previously). If these options literally always have to be the same then there's not much point in having them :grin: I'm alright if you want to leave it like this for now until we can figure out the exact right condition.

Branching STREQUAL to a separate thread since it's really a separate question

Comment (GMNGeoffrey) [inline CMakeLists.txt:131]: Branching from other thread

>> gcmn:
>> I'm also a bit worried about `STREQUAL` here since in stringly-typed CMake there are many strings that mean "true" (`1`, `ON`, `YES`, `TRUE`, `Y`, or a non-zero number) "false" (`0`, `OFF`, `NO`, `FALSE`, `N`, `IGNORE`, `NOTFOUND`, the empty string, or ends in the suffix `-NOTFOUND`) :nauseated_face: (from https://cmake.org/cmake/help/latest/command/if.html#basic-expressions). Maybe we really do want to be stricter here, but something to consider

> bjacob:
> do you mean I should be able to use `EQUAL` here ?  I dont really understand that that does, sorry. Is it "equality comparison after implict conversion to int/bool" ?

No `EQUAL` requires that the operands be numbers. When you do `if(FOO)` CMake evaluates `FOO` for a bunch of potentially truth-valued strings, not just e.g. `ON`. So if someone sets `IREE_ENABLE_TSAN=Y` and `IREE_BYTECODE_MODULE_ENABLE_TSAN=1` those are both technically true but `if(NOT IREE_ENABLE_TSAN STREQUAL IREE_BYTECODE_MODULE_ENABLE_TSAN)` will evaluate true (the error condition). The documentation for CMake `option` only talks about `ON`/`OFF` (https://cmake.org/cmake/help/v3.20/command/option.html), but AFAICT that's not actually enforced in any way (I just tried setting an option to `FOO` and it worked fine). I think the only way to force the boolean conversion would be to use `if` itself. Something like

```cmake
if(IREE_ENABLE_TSAN)
  set(IREE_ENABLE_TSAN ON)
else()
  set(IREE_ENABLE_TSAN OFF)
endif()

if(IREE_BYTECODE_MODULE_ENABLE_TSAN)
  set(IREE_BYTECODE_MODULE_ENABLE_TSAN ON)
else()
  set(IREE_BYTECODE_MODULE_ENABLE_TSAN OFF)
endif()
```

or

```cmake
if ((IREE_ENABLE_TSAN AND NOT IREE_BYTECODE_MODULE_ENABLE_TSAN) OR (IREE_BYTECODE_MODULE_ENABLE_TSAN AND NOT IREE_ENABLE_TSAN))
  # error
endif()
```

I'm not sure it's really worth it, but something to note. Ugh CMake :vomiting_face:

Comment (bjacob) [reply CMakeLists.txt:131]: OK, thanks for explaining. It's what I had feared -- and why I stuck to `STREQUAL`. I understand the concern about overly relying on the particular string truth-values, but like you say, the alternatives are not ideal either. Doing

```
if(IREE_ENABLE_TSAN)
  set(IREE_ENABLE_TSAN ON)
else()
  set(IREE_ENABLE_TSAN OFF)
endif()
```

for every option would be verbose. Doing it only for the few options that we happen to rely on that for at the moment, would be brittle.

Doing

```
if ((IREE_ENABLE_TSAN AND NOT IREE_BYTECODE_MODULE_ENABLE_TSAN) OR (IREE_BYTECODE_MODULE_ENABLE_TSAN AND NOT IREE_ENABLE_TSAN))
  # error
endif()
```

would be verbose and easy for logic bugs to sneak into.  `STREQUAL` still looks reasonable by comparison...

[the remark this answers is not in the document]
Comment (bjacob) [reply CMakeLists.txt:131]: > If these options literally always have to be the same then there's not much point in having them

I guess there's 2 reasons to have them:
* For the sake of "options that do one thing each". I have in the past proposed options doing multiple things and that was turned down on that basis :slightly_smiling_face: 
* In case we would some day figure a clear condition for "Are we building and then executing any iree_bytecode_module" so we could reenclose that if into an if on that condition, like we had before with that `if(IREE_BUILD_TESTS)` that was dropped.  In that sense, these options don't quite "literally always have to be the same", it may be just a temporary state of things.

[the remark this answers is not in the document]
Comment (GMNGeoffrey) [reply CMakeLists.txt:131]: Yeah, the latter thing is what I was referring to. SGTM :-)

Comment (GMNGeoffrey) [reply CMakeLists.txt:131]: Yeah, more just something to be aware of. I agree that if we were to do option 1, we should really do it for all our options. Can you just add a comment here about that deliberate tradeoff to stick to the simpler STREQUAL even though it doesn't quite match with CMake's absurdity

Comment (bjacob) [reply CMakeLists.txt:134]: Done

File: build_tools/cmake/iree_bytecode_module.cmake [hunk 8]
```diff
@@ -120,6 +120,13 @@ function(iree_bytecode_module)
     list(APPEND _ARGS "-iree-llvm-link-embedded=false")
   endif()
 
+  # Support testing in TSan build dirs. Unlike other sanitizers, TSan is an
+  # ABI break: when the host code is built with TSan, the module must be too,
+  # otherwise we get crashes calling module code.
+  if (IREE_BYTECODE_MODULE_ENABLE_TSAN)
```
Comment (ScottTodd) [inline build_tools/cmake/iree_bytecode_module.cmake:126]: CMake style nit: we don't put a space between `if` and `(`

**Review by ScottTodd — APPROVED** [review]

File: build_tools/cmake/iree_bytecode_module.cmake (hunk 8 above)
Comment (bjacob) [reply build_tools/cmake/iree_bytecode_module.cmake]: Done
